src/geoarrow.cpp

```diff
@@ -71,7 +71,178 @@ py::array_t<PyObjectGeography> from_geoarrow(py::object input,
+
+        ArrowArray* c_array = static_cast<ArrowArray*>(malloc(sizeof(ArrowArray)));
+        ArrowSchema* c_schema = static_cast<ArrowSchema*>(malloc(sizeof(ArrowSchema)));
+        move(&array_, &schema_, c_array, c_schema);
+
+        constexpr auto array_cleanup = [](void* ptr) noexcept {
+            auto array = static_cast<ArrowArray*>(ptr);
+            if (array->release != nullptr) {
+                array->release(array);
+            }
+            free(array);
+        };
+        py::capsule array_capsule{c_array, "arrow_array", array_cleanup};
+
+        constexpr auto schema_cleanup = [](void* ptr) noexcept {
+            auto schema = static_cast<ArrowSchema*>(ptr);
+            if (schema->release != nullptr) {
+                schema->release(schema);
+            }
+            free(schema);
+        };
+        py::capsule schema_capsule{c_schema, "arrow_schema", schema_cleanup};
+
+        return py::make_tuple(schema_capsule, array_capsule);
+    }
+
+    void move(ArrowArray* array_src,
+              ArrowSchema* schema_src,
+              ArrowArray* array_dst,
+              ArrowSchema* schema_dst) {
+        memcpy(array_dst, array_src, sizeof(struct ArrowArray));
+        array_src->release = nullptr;
+
+        memcpy(schema_dst, schema_src, sizeof(struct ArrowSchema));
+        schema_src->release = nullptr;
+    }
+
+    /// \brief Call data's release callback if valid
+    void reset() {
+        if (array_.release != nullptr) {
+            array_.release(&array_);
+        }
+
+        if (schema_.release != nullptr) {
+            schema_.release(&schema_);
+        }
+    }
+
+    /// \brief Call data's release callback if valid and move ownership of the data
+    /// pointed to by data
+    void reset(ArrowArray* array_src, ArrowSchema* schema_src) {
+        reset();
+        move(array_src, schema_src, &array_, &schema_);
+    }
+
+    ~ArrowArrayHolder() {
+        reset();
+    }
+
+protected:
+    ArrowArray array_;
+    ArrowSchema schema_;
+};
+
+ArrowArrayHolder to_geoarrow(py::array_t<PyObjectGeography> input,
+                             py::object output_schema,
+                             //  py::object geometry_encoding,
+                             bool planar,
+                             float tessellate_tolerance,
+                             int precision) {
+    ArrowArrayHolder array = ArrowArrayHolder();
+
+    s2geog::geoarrow::Writer writer;
+    std::vector<std::unique_ptr<s2geog::Geography>> s2geog_vec;
+
+    s2geog::geoarrow::ExportOptions options;
+    options.set_precision(precision);
+    if (planar) {
+        auto tol = S1Angle::Radians(tessellate_tolerance / EARTH_RADIUS_METERS);
+        options.set_tessellate_tolerance(tol);
+        // TODO make this configurable
+        // options.set_projection(s2geog::geoarrow::mercator());
+    }
+
+    if (!output_schema.is(py::none())) {
+        py::capsule schema_capsule = output_schema.attr("__arrow_c_schema__")();
```

Comment:
I guess if `output_schema` is neither None nor has `__arrow_c_schema__` defined this will raise an AttributeError here? Is that fine or should we raise a more informative error in that case?

Reply:
I added a more informative error message in the from_geoarrow code, so can do that here as well